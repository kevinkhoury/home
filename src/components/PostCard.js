import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'gatsby'
import { useInView } from 'react-intersection-observer'
import moment from 'moment'
import InlineBackgroundImage from './InlineBackgroundImage'
import { featuredImagePropTypes } from '../proptypes'

const Moment = moment().constructor

const PostCard = ({ count, image, slug, pageTitle, date }) => {
  const [ref, inView] = useInView({ triggerOnce: true })
  const isLarge = count % 3 === 0
  const classNames = [
    'post',
    'post-card',
    isLarge ? `post-card-large` : null,
    !!image ? `with-image` : `no-image`,
    inView ? 'in-view visible' : 'in-view',
  ]
    .filter((x) => x)
    .join(' ')

  const innerContent = (
    <div className="post-card-content">
      <div className="post-card-text">
        <h2 className="post-card-title">{pageTitle}</h2>
        {!!date && (
          <time dateTime={date.format('YYYY-MM-DD')} className="post-card-date">
            {date.format('MMM D, YYYY')}
          </time>
        )}
      </div>
    </div>
  )

  const content = !!slug ? (
    <Link to={slug} className="post-card-link">
      {innerContent}
    </Link>
  ) : (
    innerContent
  )

  const sources = !image
    ? null
    : !isLarge
    ? image.m.childImageSharp.fluid
    : [
        image.m.childImageSharp.fluid,
        {
          ...image.d.childImageSharp.fluid,
          media: `(min-width: 700px)`,
        },
      ]

  return (
    <article ref={ref} className={classNames}>
      {!!image ? (
        <InlineBackgroundImage fluid={sources} alt={image.alt}>
          {content}
        </InlineBackgroundImage>
      ) : (
        content
      )}
    </article>
  )
}

export const postPropTypes = {
  count: PropTypes.number.isRequired,
  image: featuredImagePropTypes,
  slug: PropTypes.string,
  pageTitle: PropTypes.string.isRequired,
  date: PropTypes.instanceOf(Moment),
}

PostCard.propTypes = postPropTypes

export default PostCard
