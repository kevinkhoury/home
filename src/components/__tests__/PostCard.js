import React from 'react'
import renderer from 'react-test-renderer'
import { render, screen } from '@testing-library/react'
import moment from 'moment'
import PostCard from '../PostCard'
import { mockAllIsIntersecting } from 'react-intersection-observer/test-utils'

const DEFAULT_PROPS = {
  count: 0,
  image: {
    src: {
      childImageSharp: {
        fluid: {},
        original: {
          height: 100,
          width: 200,
        },
      },
    },
    d: {
      childImageSharp: {
        fluid: {
          foobar: 'desktop fluid',
        },
      },
    },
    m: {
      childImageSharp: { fluid: { foobar: 'mobile fluid' } },
    },
    alt: 'imgalt',
    caption: 'imgcaption',
  },
  slug: '/foobar',
  pageTitle: 'pagetitle',
  date: moment('20200101', 'YYYYMMDD'),
}

jest.mock('../InlineBackgroundImage', () => (props) => {
  const safeProps = { ...props, children: undefined }
  return (
    <div data-testid="bg-image" className="bg-image">
      {JSON.stringify(safeProps)}
      {props.children}
    </div>
  )
})

afterEach(() => {
  jest.clearAllMocks()
})

describe('PostCard', () => {
  it('snapshot renders properly', () => {
    const tree = renderer.create(<PostCard {...DEFAULT_PROPS} />)
    expect(tree.toJSON()).toMatchSnapshot()
  })
  it('renders an image and slug only if provided', () => {
    const minimalProps = { count: 3, pageTitle: 'No posts yet' }
    render(
      <div>
        <div data-testid="full">
          <PostCard {...DEFAULT_PROPS} />
        </div>
        <div data-testid="minimal">
          <PostCard {...minimalProps} />
        </div>
      </div>,
    )
    mockAllIsIntersecting(true)
    const full = screen.getByTestId('full')
    const minimal = screen.getByTestId('minimal')
    expect(full.querySelector('.post-card-date')).toBeTruthy()
    expect(full.querySelector('.post-card-link')).toBeTruthy()
    expect(full.querySelector('.bg-image')).toBeTruthy()
    expect(minimal.querySelector('.post-card-date')).toBeNull()
    expect(minimal.querySelector('.post-card-link')).toBeNull()
    expect(minimal.querySelector('.bg-image')).toBeNull()
  })
  it('renders different container classes and images for every 4th card', () => {
    const smallProps = { ...DEFAULT_PROPS, count: 1 }
    render(
      <div>
        <div data-testid="large">
          <PostCard {...DEFAULT_PROPS} />
        </div>
        <div data-testid="small">
          <PostCard {...smallProps} />
        </div>
      </div>,
    )
    mockAllIsIntersecting(true)
    const small = screen.getByTestId('small').querySelector('.post')
    const large = screen.getByTestId('large').querySelector('.post')
    expect(small.classList.contains('post-card-large')).not.toBeTruthy()
    expect(large.classList.contains('post-card-large')).toBeTruthy()
    expect(small.classList.contains('with-image')).toBeTruthy()
    expect(large.classList.contains('with-image')).toBeTruthy()
    expect(small.querySelector('.bg-image').innerHTML).not.toMatch(
      /desktop fluid/,
    )
    expect(large.querySelector('.bg-image').innerHTML).toMatch(/desktop fluid/)
  })
})
