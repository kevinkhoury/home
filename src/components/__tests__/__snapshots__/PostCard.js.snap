// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostCard snapshot renders properly 1`] = `
<article
  className="post post-card post-card-large with-image in-view"
>
  <div
    className="bg-image"
    data-testid="bg-image"
  >
    {"fluid":[{"foobar":"mobile fluid"},{"foobar":"desktop fluid","media":"(min-width: 700px)"}],"alt":"imgalt"}
    <a
      className="post-card-link"
      href="/foobar"
    >
      <div
        className="post-card-content"
      >
        <div
          className="post-card-text"
        >
          <h2
            className="post-card-title"
          >
            pagetitle
          </h2>
          <time
            className="post-card-date"
            dateTime="2020-01-01"
          >
            Jan 1, 2020
          </time>
        </div>
      </div>
    </a>
  </div>
</article>
`;
